crates/ruff/src/jupyter/index.rs

```diff
@@ -0,0 +1,96 @@
+use std::iter;
+use std::sync::Arc;
+
+use ruff_newlines::NewlineWithTrailingNewline;
+
+use crate::jupyter::{Cell, SourceValue};
+
+/// Jupyter Notebook indexing table
+///
+/// When we lint a jupyter notebook, we have to translate the row/column based on
+/// [`ruff_text_size::TextSize`] to jupyter notebook cell/row/column.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct JupyterIndex {
+    pub(super) inner: Arc<JupyterIndexInner>,
+}
+
+impl JupyterIndex {
+    /// Returns the cell number (1-based) for the given row (1-based).
+    pub fn get_cell(&self, row: usize) -> u32 {
+        self.inner.row_to_cell[row]
+    }
+
+    /// Returns the row number (1-based) in the cell (1-based) for the
+    /// given row (1-based).
+    pub fn get_row_in_cell(&self, row: usize) -> u32 {
+        self.inner.row_to_row_in_cell[row]
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct JupyterIndexInner {
+    /// Enter a row (1-based), get back the cell (1-based)
+    pub(super) row_to_cell: Vec<u32>,
```

Comment:
Do we always need to build the index? It seems to be a very costly operation (iterates over the whole content). 

Do we need the index during fixing or is it only used when emitting diagnostics?

Reply:
Good point. The index is only required when emitting the diagnostics, I believe we could use `OnceCell` to only compute it when we're emitting them. Then, I think it would even eliminate the need to refresh the index.